0000-overhaul-qiskit-experiments.md

```diff
@@ -0,0 +1,638 @@
+# Run executor and lock ExperimentData
+
+exp_data.block_for_results()
+# Wait until lock is released
+```
+
+This is rough design for the `CircuitExecutor`, `AnalysisExecutor` and `Executor` class.
+Implementation details can be changed.
+
+```python
+class CircuitExecutor:
+    """Job manager. This can be replaced in future with Runtime API.
+    
+    Role:
+    * A black box that takes circuits and returns results.
+    * Manage backend run options.
+    """
+    
+    def __init__(self):
+        self._options = self._default_options()
+        self._executor = ThreadPoolExecutor()
+        self.backend = None
+        
+    @classmethod
+    def _default_options(cls) -> Options:
+        return Options(
+            max_circuits = None,
+        )
+    
+    def set_options(self, **options):
+        ...
+    
+    def options(self) -> Options:
+        return self._options
+    
+    def add_circuits(
+        self,
+        circuits: List[QuantumCircuit],  # Order sensitive
+    ):
+        ...
+    
+    def run(self) -> List[Dict]:  
+        # Do job splitting based on max_circuits option
+        # Manage failed job: qiskit-experiments/#1216
+        # Return canonical result dictionary
+        ...
+
+
+class AnalysisExecutor:
+    """Analysis manager. This resolves task dependency.
+    
+    Role:
+    * Run analysis callbacks in subprocess and collect result data.
+    """
+
+    def __init__(self):
+        self._executor = ProcessPoolExecutor()
+        self._dependency = Dependency()
+    
+    def add_callback(
+        self, 
+        callback: Callable,
+        predecessor: Optional[str] = None,  # task dependency
+    ) -> str:  # Returns assigned task UID
+        ...
+        
+    def run(self) -> Tuple[List[AnalysisResultData], List[Figure]]:
+        # Run callbacks in subprocess
+        # Avoid multiprocessing when # of callbacks is less than threshold
+        ...    
+
+
+class Executor:
+    """Executor interface attached to Experiment instance.
+    
+    Role:
+    * Control entire execution
+    * Bridge between experiment data and experiment class
+    """
+    
+    def __init__(self):        
+        self.experiment_data = None
+        self.circuit_executor = CircuitExecutor()
+        self.analysis_executor = AnalysisExecutor()
+    
+    def initialize_experiment_data(self):
+        ...
+    
+    def run(self):
+        # 1. Lock self.experiment_data
+        # 2. Run Job Executor, wait for results
+        # 3. Run Analysis Executor, wait for results
+        # 4. Release lock of self.experiment_data
+        ...
+
+```
+
+It's noteworthy that one can also run experiment with transpiled circuits by directly using executor interface.
+See [Qiskit-Experiments/#1222](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1222) for background.
+
+```python
+exc = Executor()
+exc.circuit_executor.backend = some_backend
+exc.circuit_executor.add_circuits(transpiled_circuits)
```

Comment:
I like this usage pattern. Where would the circuit validation hook be?

Reply:
Could you please elaborate? This is something to validate the circuit metadata against analysis class?

Reply:
Yes, I was thinking something to validate that the metadata and circuit output formats are what's expected by the analysis class. We don't necessarily have to implement it for every analysis class, but having a template hook would be useful.

Reply:
Added description in [b83b848](https://github.com/Qiskit/RFCs/pull/47/commits/b83b8483febb424c1ba53267fcf6c501f89255f7). Validation for composite experiment circuit might become too complicated to implement. Probably we need to simplify the metadata structure.